cookbook/controller/argument_value_resolver.rst

```diff
@@ -0,0 +1,163 @@
+.. index::
+   single: Controller; Argument Value Resolvers
+
+Extending Action Argument Resolving
+===================================
+
+.. versionadded:: 3.1
+    The ``ArgumentResolver`` and value resolvers are added in Symfony 3.1.
+
+In the book, you've learned that you can add the :class:`Symfony\\Component\\HttpFoundation\\Request`
+as action argument and it will be injected into the method. This is done via the
+:class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentResolver`. The ``ArgumentResolver`` uses
+several value resolvers which allow you to extend the functionality.
+
+
+Functionality Shipped With The HttpKernel
+-----------------------------------------
+
+Symfony ships with four value resolvers in the HttpKernel:
+  * The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\ArgumentFromAttributeResolver`
+    attempts to find a request attribute that matches the name of the argument.
+
+  * The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\RequestValueResolver`
+    injects the current ``Request`` if type-hinted with ``Request``, or a sub-class thereof.
+
+  * The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\DefaultValueResolver`
+    will set the default value of the argument if present and the argument is optional.
+
+  * The :class:`Symfony\\Component\\HttpKernel\\Controller\\ArgumentValueResolver\\VariadicValueResolver`
+    verifies in the request if your data is an array and will add all of them to the argument list.
+    When the action is called, the last (variadic) argument will contain all the values of this array.
+
+.. note::
+
+    In older versions of Symfony this logic was all resolved within the ``ControllerResolver``. The
+    old functionality is moved to the ``LegacyArgumentResolver``, which contains the previously
+    used resolving logic.
+
+Adding a New Value Resolver
+---------------------------
+
+Adding a new value resolver requires one class and one service defintion. In our next example, we
+will be creating a shortcut to inject the ``User`` object from our security. Given you write the following
+action::
+
+    namespace AppBundle\Controller;
+
+    use AppBundle\User;
```

Comment:
I prefer consistency and familiarity for the developer, so I don't really mind either way. I'll update this to be in the Entity namespace.